src/data/fin.lean

```diff
@@ -1012,6 +1012,168 @@ def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n
   then u ⟨i, h⟩
   else v ⟨(i : ℕ) - m, (nat.sub_lt_left_iff_lt_add (le_of_not_lt h)).2 (ho ▸ i.property)⟩
 
+variables {ι : Type*} {o : ℕ} {u : fin m → ι} {v : fin n → ι}
+
+lemma append_apply_fst (ho : o = m + n) (k : fin o) (h : (k : ℕ) < m) :
+  append ho u v k = u (k.cast_lt h) :=
+dif_pos h
+
+@[simp] lemma append_apply_fst_cast (ho : o = m + n) (k : fin m) :
+  append ho u v (cast ho.symm (cast_add n k)) = u k :=
+(append_apply_fst ho (cast ho.symm (cast_add n k)) k.2).trans $
+  congr_arg _ (fin.ext rfl)
+
+lemma append_apply_snd (ho : o = m + n) (k : fin o) (h : ¬(k : ℕ) < m) :
+  append ho u v k = v ⟨(k : ℕ) - m, (nat.sub_lt_left_iff_lt_add $ le_of_not_lt h).2 (ho ▸ k.2)⟩ :=
+dif_neg h
+
+@[simp] lemma append_apply_snd_cast (ho : o = m + n) (k : fin n) :
+  append ho u v (cast ho.symm (nat_add m k)) = v k :=
+(append_apply_snd ho (cast ho.symm (nat_add m k)) (not_lt.2 $ le_add_right (le_refl _))).trans $
+  congr_arg _ (fin.ext $ nat.add_sub_cancel_left _ _)
```

Comment:
`fst` and `snd` suggest `prod` is involved to me. I'd be inclined to either name these `append_apply_left` and `append_apply_right`, or `append_apply_of_lt`, `append_apply_of_not_lt`.